executor/src/main/java/io/kestra/executor/handler/ExecutionEventMessageHandler.java

```diff
@@ -0,0 +1,278 @@
+import io.kestra.core.services.WorkerGroupService;
+import io.kestra.core.trace.Tracer;
+import io.kestra.core.trace.TracerFactory;
+import io.kestra.core.utils.ListUtils;
+import io.kestra.core.utils.TruthUtils;
+import io.kestra.executor.*;
+import io.kestra.plugin.core.flow.WorkingDirectory;
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.api.trace.StatusCode;
+import jakarta.inject.Inject;
+import jakarta.inject.Named;
+import jakarta.inject.Singleton;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.tuple.Pair;
+import org.slf4j.event.Level;
+
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+import static io.kestra.core.utils.Rethrow.throwConsumer;
+import static io.kestra.core.utils.Rethrow.throwFunction;
+
+@Singleton
+@Slf4j
+public class ExecutionEventMessageHandler implements ExecutorMessageHandler<ExecutionEvent> {
+    @Inject
+    private ExecutionStateStore executionStateStore;
+    @Inject
+    private ExecutionQueuedStateStore executionQueuedStateStore;
+    @Inject
+    private ExecutionDelayStateStore executionDelayStateStore;
+    @Inject
+    private SLAMonitorStateStore  slaMonitorStateStore;
+    @Inject
+    private ConcurrencyLimitStateStore concurrencyLimitStateStore;
+
+    @Inject
+    private ExecutorService executorService;
+    @Inject
+    private WorkerGroupService workerGroupService;
+
+    @Inject
+    private FlowMetaStoreInterface flowMetaStore;
+
+    @Inject
+    @Named(QueueFactoryInterface.WORKERJOB_NAMED)
+    private QueueInterface<WorkerJob> workerJobQueue;
+    @Inject
+    @Named(QueueFactoryInterface.SUBFLOWEXECUTIONRESULT_NAMED)
+    private QueueInterface<SubflowExecutionResult> subflowExecutionResultQueue;
+    @Inject
+    @Named(QueueFactoryInterface.EXECUTION_NAMED)
+    private QueueInterface<Execution> executionQueue;
+    @Inject
+    @Named(QueueFactoryInterface.WORKERTASKLOG_NAMED)
+    private QueueInterface<LogEntry> logQueue;
+
+    private final Tracer tracer;
+
+    @Inject
+    public ExecutionEventMessageHandler(TracerFactory tracerFactory) {
+        this.tracer = tracerFactory.getTracer(DefaultExecutor.class, "EXECUTOR");
+    }
+
+    @Override
+    public Optional<ExecutorContext> handle(ExecutionEvent message) {
+        return executionStateStore.lock(message.executionId(), execution -> tracer.inCurrentContext(
+            execution,
+            FlowId.uidWithoutRevision(execution),
+            () -> {
+                try {
+                    final FlowWithSource flow = flowMetaStore.findByExecutionThenInjectDefaults(execution).orElseThrow();
+                    ExecutorContext executor = new ExecutorContext(execution, flow);
+
+                    // schedule it for later if needed
+                    if (execution.getState().getCurrent() == State.Type.CREATED && execution.getScheduleDate() != null && execution.getScheduleDate().isAfter(Instant.now())) {
```

Comment:
Also, to improve readability and facilitate maintenance, it might be useful to create a `var` with an explicit name for such conditition, e.g.: `isCreatedButDelayedExecution`. Same comment for other conditions below.